src/dist/replication/lib/replica_bulk_load.cpp

```diff
@@ -194,6 +194,160 @@ error_code replica::do_bulk_load(const std::string &app_name,
+        derror_replica("parse bulk load metadata failed, error = {}", err.to_string());
+        return err;
+    }
+
+    // download sst files asynchronously
+    for (const auto &f_meta : _bulk_load_context._metadata.files) {
+        auto bulk_load_download_task = tasking::enqueue(
+            LPC_BACKGROUND_BULK_LOAD, &_tracker, [this, remote_dir, local_dir, f_meta, fs]() {
+                error_code ec = ERR_OK;
+                uint64_t f_size = 0;
+                do_download(remote_dir, local_dir, f_meta.name, fs, ec, f_size);
+                if (ec == ERR_OK && !verify_sst_files(f_meta, local_dir)) {
+                    ec = ERR_CORRUPTION;
+                }
+                if (ec != ERR_OK) {
+                    try_decrease_bulk_load_download_count();
+                    _bulk_load_context._download_status.store(ec);
+                    derror_replica(
+                        "failed to download file({}), error = {}", f_meta.name, ec.to_string());
+                    // TODO(heyuchen): add perf-counter
+                    return;
+                }
+                // download file succeed, update progress
+                update_bulk_load_download_progress(f_size, f_meta.name);
+                // TODO(heyuchen): add perf-counter
+            });
+        _bulk_load_context._download_task[f_meta.name] = bulk_load_download_task;
+    }
+    return err;
+}
+
+// ThreadPool: THREAD_POOL_REPLICATION, THREAD_POOL_REPLICATION_LONG
+void replica::do_download(const std::string &remote_dir,
+                          const std::string &local_dir,
+                          const std::string &file_name,
+                          dist::block_service::block_filesystem *fs,
+                          /*out*/ error_code &download_err,
+                          /*out*/ uint64_t &download_file_size)
+{
+    // TODO(heyuchen): TBD
+    // download files from remote provider
+    // this function can also be used in restore
+}
+
+// ThreadPool: THREAD_POOL_REPLICATION
+error_code replica::parse_bulk_load_metadata(const std::string &fname,
+                                             /*out*/ bulk_load_metadata &meta)
+{
+    // TODO(heyuchen): TBD
+    // read file and parse file content as bulk_load_metadata
+    return ERR_OK;
+}
+
+// ThreadPool: THREAD_POOL_REPLICATION_LONG
+bool replica::verify_sst_files(const file_meta &f_meta, const std::string &local_dir)
+{
+    // TODO(heyuchen): TBD
+    // compare sst file metadata calculated by file and parsed by metadata
+    return true;
+}
+
+// ThreadPool: THREAD_POOL_REPLICATION_LONG
+void replica::update_bulk_load_download_progress(uint64_t file_size, const std::string &file_name)
+{
+    // TODO(heyuchen): TBD
+    // update download progress after downloading sst files succeed
+}
+
+// ThreadPool: THREAD_POOL_REPLICATION, THREAD_POOL_REPLICATION_LONG
+void replica::try_decrease_bulk_load_download_count()
+{
+    if (_stub->_bulk_load_downloading_count.load() > 0) {
```

Comment:
There maybe a  multithread problem, suppose _bulk_load_downloading_count == 1, and there are 2 threads consult `if (_stub->_bulk_load_downloading_count.load() > 0)`, both of them is true, them both of them fetch_sub(1), the result is -1.

Reply:
I think the result of the multi-thread problem can be accepted. If the result is -1, it may lead actual downloading count = restriction count + 1. However, if you think it is necessary to avoid such problem, I can add a lock for it.

Reply:
Use atomic `++` and `--` instead, and add debug log here